Show free shipping indication only if appropriate

hacchus opened the pull request:
The problem is if you are not logged in or need postal codes to determine shipping, you always see shipping as "Free" in your cart, which really means "not calculated yet". It is bad for customer psychology to see "Free" and after they enter more information to get a price.

The result of this change is if there is shipping calculated then it will simply show it. If the system cannot determine shipping yet for any reason then you get "Not Calculated". Finally, if you really have free shipping then you get the indication "Free".


| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?       | Instead of showing "Free", show an indication in cart like "Not Calculated" for shipping if shipping cost cannot yet be determined.
| Type?         |  improvement
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #10739 #15112

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/13517)
<!-- Reviewable:end -->

hacchus committed:
Show free shipping indication only if appropriate

The problem is if you are not logged in or need postal codes to determine shipping, you always see shipping as "Free" in your cart, which really means "not calculated yet". It is bad for customer psychology to see "Free" and after they enter more information to get a price.

The result of this change is if there is shipping calculated then it will simply show it. If the system cannot determine shipping yet for any reason then you get "Not Calculated". Finally, if you really have free shipping then you get the indication "Free".

LouiseBonnard commented on src/Adapter/Presenter/Cart/CartPresenter.php:
```diff
@@ -475,6 +473,41 @@ function ($voucher) {
+        if ($shippingCost != 0) {
+            $shippingDisplayValue = $this->priceFormatter->format($shippingCost);
+        } else if ($hasFreeCarrierFirst == 0) {
+            $shippingDisplayValue = $this->translator->trans('Not Calculated', array(), 'Shop.Theme.Checkout');
```
Hey, thank you, wouldn't it be easier to remove this line when shipping costs are not determined yet and display it only when free shipping is defined in parameters? If not, I would suggest _Calculated later_ or _Calculated at checkout_. Ping me if you all have a better idea!

hacchus replied:
@LouiseBonnard Yes"Calculated later" seems better. If anyone else has a better idea let's hear it.

rblaurin replied:
I would vote for the first suggestion from @LouiseBonnard, removing this line when not calculated...

hacchus replied:
@marionf @rblaurin @LouiseBonnard 

Yes we can hide it. I think we need to change the theme in two places (if more please let me know).

First we send an empty string instead of "Not Calculated".

For modal cart we must change this file:
themes\classic\modules\ps_shoppingcart\modal.tpl on line 60 and surround it with an if statement to check "$cart.subtotals.shipping.value"

For checkout page this file: themes\classic\templates\checkout\_partials\cart-summary.tpl on line 56 we change "if $subtotal" with "if $subtotal.value"

The file themes\classic\templates\checkout\_partials\cart-detailed-totals.tpl is already doing this on line 30 so we don't need to change anything.

I tested this on 1.7.5.2 and works fine. Are there other places we need to change?

hacchus committed:
Removed extra condition and minor code changes

Removed condition for $currentCarrier['position'] == 0 because it's not necessary and can cause issues for different back-end carrier options.

marionf commented:
@hacchus 

> The problem is if you are not logged in or need postal codes to determine shipping, you always see shipping as "Free" in your cart, which really means "not calculated yet". 

The price of the carrier that you see when you are not logged depends of the default carrier and default country.
You see "free" if for your default country and default carrier you have defined no amount
If you choose as default a not free carrier, you will see price even if you are not connected

![capture d'écran_1485](https://user-images.githubusercontent.com/13449658/56963227-c3560780-6b58-11e9-88a6-3314765551aa.png)


I just tried your PR and disable all carriers, I still see "free" shipping

hacchus commented:
@marionf 

I've tested this on a fresh 1.7.5.2 installation and if I disable all carriers then I get the "Not Calculated" indication as expected. I also have a clean 1.7.5.0 installation with the same results.

![dc1](https://user-images.githubusercontent.com/5149317/57104846-0f0fd900-6d32-11e9-8f03-f1d550a8bd6a.jpg)

Most of the times I use 3rd party modules to display shipping with complex rules that require zip codes, and therefore it's not possible to calculate shipping until I have zip codes. Up to that point I always get "Free" which is misleading, that's why I came up with this solution. 

I made this modification on two shops (1.7.4.4 and 1.7.5.1), that use a 3rd party shipping module with zip codes and it works as expected.

marionf commented:
@hacchus 

Thank you for your answer, my bad I was testing your PR on 1.7.6.x branch instead of develop branch
I well see "not calculated" now.
Do you think it's possible to not display this line instead of "not calculated" ?
I am not sure people understand why it's "not calculated"

hacchus committed:
Hide shipping line if shipping can’t be determined

marionf commented:
Thank you @hacchus 
Let's wait for code review now

matks commented on themes/classic/templates/checkout/_partials/cart-summary.tpl:
```diff
@@ -54,7 +54,7 @@
     {block name='cart_summary_subtotals'}
       {foreach from=$cart.subtotals item="subtotal"}
-        {if $subtotal && $subtotal.type !== 'tax'}
+        {if $subtotal.value && $subtotal.type !== 'tax'}
```
This line might have side-effects. If `$subtotal.value` is 0 (integer), it will hide the line instead of display it. This could create issue if the merchant wants to display some subtotal to the customer for legal compliance matters 🤔 ?

matks replied:
I say this because the change will impact every subtotals passed in `$cart.subtotals`, not only the shipping subtotal 😉

matks commented on src/Adapter/Presenter/Cart/CartPresenter.php:
```diff
@@ -475,6 +473,39 @@ function ($voucher) {
+    private function getShippingDisplayValue($cart, $shippingCost)
+    {
+        $hasFreeCarrier = 0;
+        $default_country = null;
```
Can we stick to camelCase variable naming 😉 ? Just above we have `$hasFreeCarrier` that is camelCase then this one `$default_country` that is snake_case

matks commented on src/Adapter/Presenter/Cart/CartPresenter.php:
```diff
@@ -475,6 +473,39 @@ function ($voucher) {
         );
     }
+    
+    private function getShippingDisplayValue($cart, $shippingCost)
```
Can we have phpDoc please ?

matks commented on src/Adapter/Presenter/Cart/CartPresenter.php:
```diff
@@ -475,6 +473,39 @@ function ($voucher) {
+            $default_country = new Country(Context::getContext()->cookie->id_country);
+        }
+
+        $delivery_option_list = $cart->getDeliveryOptionList($default_country);
```
Same, camelCase for the ❤️

matks commented on src/Adapter/Presenter/Cart/CartPresenter.php:
```diff
@@ -475,6 +473,39 @@ function ($voucher) {
+        $default_country = null;
+
+        if (isset(Context::getContext()->cookie->id_country)) {
+            $default_country = new Country(Context::getContext()->cookie->id_country);
```
This could be improved but it's a bit outside of this PR scope. If we cannot find the country of the visitor because he has no cookies but he has landed on the EN version of the website or the GB version of the website, we could still check for the relevant carriers. But this is a bonus, not mandatory.

matks commented:
Hi @hacchus thanks for the contribution ! I have a few comments, but the important one is https://github.com/PrestaShop/PrestaShop/pull/13517#discussion_r282508618. Other ones are rather for code style.

hacchus committed:
Allow subtotals 0 integer values to go through and code style fixes

hacchus commented:
Hello @matks thanks for your feedback. I did the code style changes and also moved the code that checks carriers inside the if statement, so it doesn't have to run every time, only when shipping cost is 0.

I had to also change the file themes\classic\templates\checkout_partials\cart-detailed-totals.tpl on line 30 to allow 0 values to pass from there as well.

matks commented on themes/classic/templates/checkout/_partials/cart-summary.tpl:
```diff
@@ -54,7 +54,7 @@
     {block name='cart_summary_subtotals'}
       {foreach from=$cart.subtotals item="subtotal"}
-        {if $subtotal && $subtotal.type !== 'tax'}
+        {if $subtotal && $subtotal.value|count_characters > 0 && $subtotal.type !== 'tax'}
```
Smarty way to handle not-empty values 😄 👏

kpodemski commented:
ping ping ping, is there any way to rush it to 1.7.6? this is a great contribution

cc @matks

jolelievre commented:
> ping ping ping, is there any way to rush it to 1.7.6? this is a great contribution
> 
> cc @matks

hello @kpodemski,

sorry but we already froze the 176 version, so we only integrate regression bug fixes, this will have to wait for the 177 version
but thank you @hacchus for this improvement!

matks commented:
H @hacchus sorry for the late answer, summer got us very busy with 1.7.6 release

I think the PR is good as it is 👍let's move to QA !

matks commented:
@hacchus Oh, PrettyCI linter is unhappy and there are some git conflicts. Can you handle them ? Else I can do it as you waited for a long time 😄

hacchus commented:
Hi @matks could you please do it, as I'm very busy at this time? Thanks! :grin:

khouloudbelguith commented:
Hi @hacchus,

This PR needs to be rebased into develop

Thanks!

matks committed:
Apply php-cs-fixer

matks commented:
@hacchus PR has been rebased, conflicts solved, codestyle fixed 👍we can now move to the QA stage

sarahdib commented:
Hello @hacchus 

Thank you for your contribution.

After several test we didn't see any change with the behavior before and after the PR.

Can you explain the exact step to reproduce.

Thank you

sarahdib commented:
@hacchus Sorry I think it was a cache problem from me.

matks commented:
Thank you @hacchus
